File: examples/python/encode.py
# Example on how to write WARC files.
import subprocess
import zlib

import message


def main():
    # Launch the warcat program. The options provided will tell it to read
    # JSON as a line from standard in.
    # Ensure you have warcat on the search path or adjust the path as needed.
    with subprocess.Popen(
        [
            "warcat",
            "import",
            "--compression=none",
            "--format=jsonl",
        ],
        stdin=subprocess.PIPE,
    ) as process:
        # Write a record header with the given header fields.
        # Note: this header is not valid; it is simply a concise demonstration.
        header = message.Header(
            "WARC/1.1",
            [
                ("WARC-Record-Type", "resource"),
                ("Content-Length", "12"),
            ],
        )
        message.encode(process.stdin, header)

        # Write the record block data.
        checksum = 0

        data = b"Hello world!"
        checksum = zlib.crc32(data, checksum)

        block_chunk = message.BlockChunk(data)
        message.encode(process.stdin, block_chunk)

        # Write the end of the block message.
        block_end = message.BlockEnd(checksum)
        message.encode(process.stdin, block_end)

        # Finish writing the file.
        message.encode(process.stdin, message.EndOfFile())


if __name__ == "__main__":
    main()

File: examples/encode.rs
//! Example on how to encode a WARC file by records
//! use std::io::Write;

use std::io::Write;

use warcat::{
    header::WarcHeader,
    warc::{Encoder, EncoderConfig},
};

fn main() -> anyhow::Result<()> {
    // For this example, our file is just a in-memory buffer
    let mut warc_file = Vec::new();

    // Configure the compression format if needed, otherwise use default
    let config = EncoderConfig::default();

    // Create a new WARC encoder
    let mut encoder = Encoder::new(&mut warc_file, config);

    // Write a header of a WARC record and return a block encoder.
    // Note that `write_header()` consumes the encoder and returns a
    // decoder of a different type. This is known as the typestate pattern.
    let header = WarcHeader::new(12, "Resource");
    let mut block_encoder = encoder.write_header(&header)?;

    // Write the block like a file.
    block_encoder.write_all(b"Hello world!")?;

    // Get a header encoder. Again, this is the typestate pattern.
    encoder = block_encoder.finish_block()?;

    // Get the inner writer if needed
    let _file = encoder.finish()?;

    println!("Wrote {} bytes", warc_file.len());

    Ok(())
}
